### PR TITLE
added BulkFileOperation listener so that we can pause our ambient wor…

### PR DESCRIPTION
**Pull request by heejaechang**
…k while more important operation is running

### Customer scenario

Customer switches git branch and it feels a little bit more snappier.

### Bugs this fixes

N/A

### Workarounds, if any

No workaround

### Risk

Low. it won't cause VS crash or code loss.

### Performance impact

It will stop our ambient features (LUT, Diagnostics, cache builder, source based discovery and etc) from running so user explicit action such as git switch can have more machine resources.

### Is this a regression from a previous update?

No

### Root cause analysis

as user switch git branches, a lot of files on disk will change and VS will pick them up and start to recalculate/run ambient features. problems are, it might repeatedly do so since it can't predict when is actually last change. this cause outer platform (project system) explicitly tell us when it starts and ends so that we can explicitly pause until all operation is done.

### How was the bug found?

Ongoing, perf investigation on branch switching and solution reload

### CODE REVIEW

**Comment by heejaechang** [comment]
this will fail until this checks in - https://github.com/dotnet/roslyn/pull/27027

**Comment by heejaechang** [comment]
tagging @jinujoseph @dotnet/roslyn-ide @JieCarolHu @ivanbasov @dpoeschl can you take a look?

**Comment by heejaechang** [comment]
tagging @PavelUstinovMS can you take a look?

File: src/VisualStudio/Core/Def/RoslynPackage.cs [hunk 1]
```diff
@@ -208,5 +211,71 @@ private void UnregisterRuleSetEventHandler()
...
+            // so we only need simple tracking.
+            object gate = new object();
+            GlobalOperationRegistration localRegistration = null;
+
+            BulkFileOperation.Begin += (s, a) =>
+            {
+                StartBulkFileOperationNotification();
+            };
+
+            BulkFileOperation.End += (s, a) =>
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/RoslynPackage.cs]: What is the threading expectation here -- we can get bulk operations from any thread? What about icky cases like:

1. Register for BulkFileOperation.Begin
1. An operation starts, so we start the global operation.
1. An operation ends, but we haven't subscribed yet.
1. Register for BulkFileOperation.End.

File: src/VisualStudio/Core/Def/RoslynPackage.cs [hunk 2]
```diff
@@ -208,5 +211,71 @@ private void UnregisterRuleSetEventHandler()
...
+            {
+                StopBulkFileOperationNotification();
+            };
+
+            if (BulkFileOperation.IsInProgress)
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/RoslynPackage.cs]: Technically also icky cases where IsInProgress can return true, our thread yields, an operation ends, and then this starts something again.

File: src/VisualStudio/Core/Def/RoslynPackage.cs [hunk 3]
```diff
@@ -208,5 +211,71 @@ private void UnregisterRuleSetEventHandler()
...
+            // such as solution crawler, pre-emptive remote host synchronization and etc. any background work users didn't
+            // explicitly asked for.
+            //
+            // this should give all resources to BulkFileOperation. we do same for things like build, 
+            // debugging, wait dialog and etc. BulkFileOperation is used for things like git branch switching and etc.
+            var globalNotificationService = _workspace.Services.GetService<IGlobalOperationNotificationService>();
+
+            // BulkFileOperation can't have nested events. there will be ever only 1 events (Begin/End)
+            // so we only need simple tracking.
+            object gate = new object();
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/RoslynPackage.cs]: Loving this use of local functions!

**Comment by heejaechang** [comment]
@PavelUstinovMS can you provide API which we can provide both begin and end event handler and you do right subscription? right now, unless we subscribed to it very carefully, it can easily have race. for example, if I subscribe to begin first and then end later, I can miss "end" event if I got events between begin and end subscription. same goes for IsInProgress, depending on where I check, I can get into race.

it would be nice if there is API that I give both event handler, and BulkFileOperation does right thing for us.

tagging @jasonmalinowski

**Comment by sharwell** [comment]
There are two possible situations which we want to make sure and avoid:

1. Crashes. We seem to be covered here although we may want to limit the number of calls we allow to `WatsonReporter.Report`.
2. Leaving the global operation notification service in a bulk change state while the IDE is not in a bulk change state.

I believe the current implementation avoids the second failure as long as we can verify that `BulkFileOperation` linearizes the event handling (`End` is never raised before the system completes calling all `Start` handlers, and vice versa). If the event handlers are called asynchronously, this will require additional scrutiny/clarity.

**Comment by heejaechang** [comment]
already tagged pavel.

**Comment by jasonmalinowski** [comment]
Chatting with @heejaechang yesterday on this point, this now does look correct based on our understanding of the API implementation. @heejaechang's request for a newer API was largely me telling him "it'd be great if the API was a path to success instead of a pit of fail".

**Comment by PavelUstinovMS** [comment]
Looks good to me. I think right now subscribing to End first and then to Begin should solve race problem. If you miss Begin event (for the branch switch I think this is impossible, because first we load solution and then switch branch and Roslyn is initialized during the loading) this will not crash/hang system, but rather we just miss opportunity to optimize workflow in this particular case. I'd suggest to check BulkFileOperation.IsInProgress after you subscribed to events to make it more robust and not miss opportunity for optimization. Also we prevent running multiple BFO at the same time, so there won't be 2+ Begin events without End.

**Comment by jasonmalinowski** [comment]
@PavelUstinovMS @heejaechang originally did check IsInProgress, but that is only safe if there are other assumptions. Is it the case begin/end are raised on the UI thread, or only triggered from code running on the UI thread? If not, then consider the case:

1. A batch is already running.
1. Roslyn has subscribed to Begin/End, and checks IsInProgress.
1. Imagine that Roslyn's UI thread is now "paused", i.e. context switched away.
1. Some other thread completes the batch that was running in 1. Roslyn and everybody else gets notified of the End.
1. Roslyn's UI thread resumes. It thinks there is an active batch, but there isn't, and now we're suppressing this work until some later batch operation finally kicks this.

If we have stronger guarantees, then maybe we're fine checking. I couldn't find docs on what guarantees we have so wasn't sure.

**Comment by heejaechang** [comment]
rerun test so that it can pick up new test machine that has 15.8 latest installed.

**Comment by heejaechang** [comment]
require https://github.com/dotnet/roslyn/pull/27027

**Comment by heejaechang** [comment]
@jinujoseph changed it to 15.8.x

**Comment by heejaechang** [comment]
@jinujoseph sorry I am not following. so I need to target 15.8.x-vs-deps?

**Comment by dpoeschl** [comment]
Retriggering tests. The unit tests were failing but that shouldn't have anything to do with the VS version installed on the machine. Let's see the new results.

**Comment by heejaechang** [comment]
@dpoeschl tests are failing because vssdk didn't move to latest vssdk yet that contains these APIs.

**Comment by dpoeschl** [comment]
Ahh, I see your comment above now.

😭

**Comment by heejaechang** [comment]
this is in now

https://github.com/dotnet/roslyn/pull/27027

**Comment by heejaechang** [comment]
@jasonmalinowski @jmarolf @jinujoseph 

this is still not there
https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2Fenv%2Fshell%2FPackageFramework%2FCurrent%2FShell%2FBulkFileOperation%2FBulkFileOperation.cs&version=GBlab%2Fvsumain

we might need to move to even more recent vssdk?

**Comment by jinujoseph** [comment]
yes , most probably @jasonmalinowski  was planning to move to latest as a step 2 to this

**Pull request merged**